### PR TITLE
Move tools/ci_* projects to airbyte-ci, update to use Poetry, bump to python 3.10

### LINKED ISSUE #25170: Move pipelines into the airbyte-ci project
## Overview
This is about colocating the pipelines project in the airbyte-ci folder to allow for better code sharing.

## What to do
- [ ] Move pipelines and any other related projects from ci-connector ops into the airbyte-ci folder
- [ ] Look for any todos in pipeline related to refactoring logic to be imported from another library
- [ ] Update dependencies to use poetry

### LINKED ISSUE #27865: connectors-ci: clear documentation on how to install and update `airbyte-ci`
Edit this doc to explain how to install and update `airbyte-ci`

What's currently not great:
* The remote install with `@git` is not always working
* We're not mentioning that `pip` should be upgraded
* We're not explaining how to update the CLI

### PR DESCRIPTION
**Pull request by bnchrch**
## What

1. moves tools/ci_* projects to airbyte-ci
2. Updates them to use poetry/pipx
3. Updates these packages to python 3.10
4. Improves the installation story
5. Moves pipelines to its own package

Now to install airbyte-ci all you have to run is 

```
pipx install airbyte-ci/connectors/pipelines
```
## Notes for reviewer

@alafanechere I would love to get your take on this being the `airbyte-ci` tzar.

- Is this a good idea?
- Is this the direction you had wanted to go in?
- What do you think about using pipx to isolate the install?
- The folders `ci` and `connectors` at the top of `airbyte-ci` still feel awkward is there a better way?

## Left TODO
- [x] Update `ci_code_validator`
- [x] Update GHA
- [x] Update docs
- [x] Update Dagger build 
- [x] Test that everything still works

closes #25170 
closes #27865

### CODE REVIEW

File: airbyte-ci/ci/ci_credentials/README.md
Comment (bnchrch) [inline airbyte-ci/ci/ci_credentials/README.md:1]: Pipelines/airbyte-ci are now their own package

Comment (bnchrch) [inline airbyte-ci/ci/ci_credentials/README.md:1]: Pipelines/airbyte-ci are now their own package

File: airbyte-ci/ci/ci_pipelines/README.md
Comment (bnchrch) [inline airbyte-ci/ci/ci_pipelines/README.md:1]: Pipelines/airbyte-ci are now their own package

File: airbyte-ci/ci/ci_pipelines/README.md [hunk 1]
```diff
@@ -1,29 +1,43 @@
...
-cd airbyte
-airbyte-ci
+python -m pip install --user pipx
+python -m pipx ensurepath
 ```
 
+Once pipx is installed, navigate to the root directory of the project, then run:
+
+```bash
+pipx install airbyte-ci/ci/ci_pipelines/
```
Comment (bnchrch) [inline airbyte-ci/ci/ci_pipelines/README.md:35]: The installation is one line

File: airbyte-ci/ci/ci_credentials/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+[tool.poetry]
+name = "ci_credentials"
+version = "1.1.0"
+description = "CLI tooling to read and manage GSM secrets"
+authors = ["Airbyte <contact@airbyte.io>"]
+readme = "README.md"
+packages = [{ include = "ci_credentials" }]
+
+[tool.poetry.dependencies]
+python = "^3.10"
```
Comment (bnchrch) [inline airbyte-ci/ci/ci_credentials/pyproject.toml:14]: Python 3.10 all around!!!

File: airbyte-ci/ci/ci_credentials/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+requests = "^2.28.2"
+click = "^8.1.3"
+pyyaml = "^6.0"
+ci_common_utils = { path = "../ci_common_utils", develop = true }
```
Comment (bnchrch) [inline airbyte-ci/ci/ci_credentials/pyproject.toml:18]: If we go ahead with this we will need to update the local dependency resolution right?

File: airbyte-ci/ci/ci_credentials/pyproject.toml (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/ci/ci_credentials/pyproject.toml:14]: ci_credentials is installed on GHA workflows that are running `3.9`. Can we keep compatibility of `ci_credentials` with `3.9` while suggesting use of `3.10` though `.python-version` files?

**Comment by alafanechere** [comment]
> Is this a good idea?

Yes it is! 

> Is this the direction you had wanted to go in?

Totally

> What do you think about using pipx to isolate the install?

I'm not against it but I'm not familiar with it. Is it providing dependency isolation while not requiring the use of a python virtualenv? I'm 👍 , but if poetry has a builtin way to achieve the same I'd prefer to stick to poetry (I'm not too familiar with it too..)

> The folders ci and connectors at the top of airbyte-ci still feel awkward is there a better way?

So in this `ci_connector_ops`  /  `ci_credentials` / `ci_common_utils` thing we have:
* QA checks
* `airbyte-ci` pipelines
* `ci_credentials` to download GSm secrets
* `ci_common_utils` which only has a `GoogleAPI` client class... Which I believe is only used by `ci_credentials`.

My suggestion would be to: 
* Merge `ci_credentials` and `ci_common_utils` into a single package : `airbyte-ci/connectors/credentials` ?
 * I'd love to have a package we can use as a lib that would have the `Connector` utility class. `airbyte-ci/connectors/utils` ? It could be imported in `pipelines` and `qa_checks` for instance.
 * The QA checks should be a light package in `airbyte-ci/connectors/qa_checks`

File: airbyte-ci/ci/ci_credentials/pyproject.toml (hunk 2 above)
Comment (bnchrch) [reply airbyte-ci/ci/ci_credentials/pyproject.toml:14]: I believe we can!

**Comment by alafanechere** [comment]
Allright, let's use pipx!
And let's refactor `ci_common_utils and ci_credentials` in a different PR!

> What I can do is rename these to remove ci and place them all under airbyte-ci/connectors/

👍 

> I'd love to have a package we can use as a lib that would have the Connector utility class. airbyte-ci/connectors/utils ? It could be imported in pipelines and qa_checks for instance.

Do you have an opinion about this?

**Comment by bnchrch** [comment]
> > I'd love to have a package we can use as a lib that would have the Connector utility class. airbyte-ci/connectors/utils ? It could be imported in pipelines and qa_checks for instance.
> 
> Do you have an opinion about this?

I LOVE this. Thats my opinion. I think were in a great spot to do that. Also it would be useful in the metadata_service

**Comment by bnchrch** [comment]
Pre-Publish that ran successfully: https://github.com/airbytehq/airbyte/actions/runs/5651515543/job/15309740249

**Comment by bnchrch** [comment]
**GHA Connector Tests**
source-fresh-desk: https://github.com/airbytehq/airbyte/actions/runs/5651586425
source-mysql: https://github.com/airbytehq/airbyte/actions/runs/5651609514
destination-postgres: https://github.com/airbytehq/airbyte/actions/runs/5651597479

**Comment by octavia-squidington-iii** [comment]
##  source-faker test report (commit [`26884cd0db`](https://github.com/airbytehq/airbyte/pull/27957/commits/26884cd0dbe194727b5982e7b67c96e7b60573a6)) - ❌

⏲️  Total pipeline duration: 03mn30s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690249379/26884cd0dbe194727b5982e7b67c96e7b60573a6/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

File: airbyte-ci/connectors/ci_code_validator/__init__.py
Comment (bnchrch) [inline airbyte-ci/connectors/ci_code_validator/__init__.py:1]: @airbytehq/connector-operations Gut check: We should remove this.

There is
1. no use of `ci_sonar_qube` in code
2. no use of `ci_changes_detection` in code 
3. the is one use of `ci_sonar_qube` in a github action: https://github.com/airbytehq/airbyte/pull/27957/files#diff-c46f6e02d0050829e4e9f4a4bd0eb29da35853e0786b7ddf348f665085c561da
4. But that action is not used.

File: .github/workflows/test-performance-command.yml
Comment (bnchrch) [inline .github/workflows/test-performance-command.yml:1]: This should be deleted. 

See: https://github.com/airbytehq/airbyte/pull/27957/files#r1272911875

**Comment by bnchrch** [comment]
@airbytehq/connector-operations This is ready for review!

Most of this is
1. a folder restructure
2. a tooling change
3. a python version bump

The only file with net new logic is how we install pipx and use it in the pipelines here

[...tor_ops/pipelines/actions/environments.py → ...pelines/pipelines/actions/environments.py](https://github.com/airbytehq/airbyte/pull/27957/files#diff-b1c73385b3a55c0d61d11e4323076b841402c0401100fc95d902b9c4d87609fe)

Please let me know what you think, I would like this to be a speedy merge to avoid conflicts!

File: airbyte-ci/connectors/ci_code_validator/__init__.py
Comment (erohmensing) [reply airbyte-ci/connectors/ci_code_validator/__init__.py:1]: I've never heard of this, if that helps anything!

**Review by erohmensing — APPROVED** [review]
Installation was a breeze! Messed around with formatting a source, testing the orchestrator and metadata lib. I'm still not sure (via READMEs alone, and after a quick non-deep perusal of CI changes) how to run tests for the pipelines. 

Some non-blocking feedback - I think we should merge this improvement before trying to tweak anything much more: 
* I was able to run the QA engine from where it lives with the instructions provided. I find it sort of weird it lives under airbyte ci but has a separate installation and run process
* It looks like the README of airbyte-ci lives under `airbyte-ci/connectors/pipelines` - i'd expect it to live under `airbyte-ci` or maybe `airbyte-ci/connectors`. I guess this makes sense because of the install command being `pipx install airbyte-ci/connectors/pipelines`... this installation path confuses me since i expect airbyte-ci to be able to run the metadata service as well (and the qa engine as described above)

**Review by alafanechere — APPROVED** [review]
Approving to unblock but could you please checkout the report path problem and the dev mode story? 😄

File: airbyte-ci/ci/ci_credentials/pyproject.toml (hunk 2 above)
Comment (pedroslopez) [reply airbyte-ci/ci/ci_credentials/pyproject.toml]: Is there a reason to keep the version diff between local and ci? Could this be another source of behavior mismatches?

Comment (pedroslopez) [reply airbyte-ci/ci/ci_credentials/pyproject.toml]: nvm, seems like we bumped most things as we went along, except `.github/workflows/connector-performance-command.yml` for some reason?

Comment (bnchrch) [reply airbyte-ci/ci/ci_credentials/pyproject.toml]: Oh I think that was a miss on my part.

I ended up going with 3.10 everywhere as it seems that there were no issues encountered while running these tools as 3.10

File: airbyte-ci/connectors/pipelines/pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+[tool.poetry]
+name = "pipelines"
+version = "0.1.0"
+description = "Packaged maintained by the connector operations team to perform CI for connectors' pipelines"
+authors = ["Airbyte <contact@airbyte.io>"]
+
+[tool.poetry.dependencies]
+python = "^3.10"
+dagger-io = "^0.6.4"
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pyproject.toml:13]: As I understand it. This is expected!

https://github.com/pypa/pipx#how-is-it-different-from-pip

pipx actually installs all dependencies for the specified package in a directory isolated from your current python environment.

The packages you have in you global env or pyenv will have no effect on the package dependencies install with pipx and vice versa

So it makes sense to me that the dagger-io version in your pyenv shell didnt change.

But can you tell me what you ran to get the version for dagger-io?

File: airbyte-ci/connectors/pipelines/pipelines/consts.py [hunk 5]
```diff
@@ -32,6 +32,6 @@
 GRADLE_CACHE_PATH = "/root/.gradle/caches"
 GRADLE_BUILD_CACHE_PATH = f"{GRADLE_CACHE_PATH}/build-cache-1"
 GRADLE_READ_ONLY_DEPENDENCY_CACHE_PATH = "/root/gradle_dependency_cache"
-LOCAL_REPORTS_PATH_ROOT = "tools/ci_connector_ops/pipeline_reports/"
+LOCAL_REPORTS_PATH_ROOT = "airbyte-ci/connectors/pipelines/pipeline_reports/"
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/consts.py:35]: Ooh good catch. Let me see whats going on

File: airbyte-ci/connectors/pipelines/README.md [hunk 6]
```diff
@@ -1,30 +1,50 @@
...
+
 If you face any installation problem feel free to reach out the Airbyte Connectors Operations team.
-N.B: This project will eventually be moved to `airbyte-ci` root directory.
 
+### Installation for development
+If you are developing on pipelines, we recommend installing airbyte-ci in editable mode:
+
+```bash
+pipx install airbyte-ci/connectors/pipelines/ --force --editable
+```
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/README.md:47]: Thumbs up to using poetry for dev!

Seems a lot more approachable and intuitive :)

File: airbyte-ci/connectors/pipelines/README.md [hunk 7]
```diff
@@ -1,30 +1,50 @@
...
 ```bash
-# Make sure that the current Python version is >= 3.10
-pyenv shell 3.10
-pip install "ci-connector-ops[pipelines] @ git+https://github.com/airbytehq/airbyte.git@master#subdirectory=tools/ci_connector_ops"
-cd airbyte
-airbyte-ci
+pipx install airbyte-ci/connectors/pipelines/
 ```
 
+This command installs `airbyte-ci` and makes it globally available in your terminal.
```
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/README.md:38]: Could you elaborate a bit more on the why and what an ideal install instruction looks like?

File: airbyte-ci/connectors/pipelines/pipelines/consts.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bnchrch) [reply airbyte-ci/connectors/pipelines/pipelines/consts.py:35]: Ah I see whats going on

**empty pipeline_reports folder created**
This is an is issue with where we [create the report directory](https://github.com/airbytehq/airbyte/blob/370c3fb884ee3ae47684c9d595169d23ffd08037/airbyte-ci/connectors/pipelines/pipelines/utils.py#L452) which is only discovered if you run the `airbyte-ci` cmd not in the root.

Ill move this closer to us writing the log

**And the actual test reports are written a very nested location:**
It is very nested but expected

The log path `airbyte-ci/connectors/pipelines/pipeline_reports/airbyte-ci/connectors/test/manual/bnchrch_ci_poetry-pipelines/1690300788/26884cd0dbe194727b5982e7b67c96e7b60573a6/dagger.log` has three discrete parts

1. The parent folder `airbyte-ci/connectors/pipelines/`
2. The airbyte-ci defined path structure `airbyte-ci/connectors/test/`
3. the git info `manual/bnchrch_ci_poetry-pipelines/1690300788/26884cd0dbe194727b5982e7b67c96e7b60573a6`

so you will note that #1 and #2 are nearly identical. 

and that turns out is a coincidence.

The airbyte-ci defined path structure is derived from the click command hierarchy `airbyte-ci connectors test --name=...`

**Comment by octavia-squidington-iii** [comment]
##  source-faker test report (commit [`68758f99ba`](https://github.com/airbytehq/airbyte/pull/27957/commits/68758f99ba89dfbd385b70129f33cd89762aee48)) - ❌

⏲️  Total pipeline duration: 03mn22s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690335663/68758f99ba89dfbd385b70129f33cd89762aee48/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

**Comment by bnchrch** [comment]
@alafanechere @pedroslopez 

Just updated the branch to
1. Have poetry based dev install instructions
2. Have pipeline test deps declared
3. Remove ci_code_validator folder and related action
4. Upgrade `.github/workflows/connector-performance-command.yml` to 3.10
5. Ran the python formatter on airbyte-ci
6. Updated the logging code to not create the dagger.log parent folder until nessesary

Im going to run some test pipelines now to ensure everything works as planned

**Comment by octavia-squidington-iii** [comment]
##  source-faker test report (commit [`4090391bac`](https://github.com/airbytehq/airbyte/pull/27957/commits/4090391bacd54a11de7059de732f1630349361ba)) - ❌

⏲️  Total pipeline duration: 03mn17s 

| Step                                                                | Result   |
|:--------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-faker/metadata.yaml | ✅        |
| Connector version semver check                                      | ✅        |
| Connector version increment check                                   | ❌        |
| QA checks                                                           | ✅        |
| Code format checks                                                  | ✅        |
| Connector package install                                           | ✅        |
| Build source-faker docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                          | ✅        |
| Acceptance tests                                                    | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/bnchrch_ci_poetry-pipelines/1690336063/4090391bacd54a11de7059de732f1630349361ba/source-faker/4.0.0/output.html)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-faker test
```

**Comment by bnchrch** [comment]
**Test**
- [GHA run ](https://github.com/airbytehq/airbyte/actions/runs/5663728190)with args `--name=source-postgres --name=destination-postgres --name=source-fresh-desk` 

**prepublish**
- [source-postgres](https://github.com/airbytehq/airbyte/actions/runs/5663736440)
- [source-fresh-desk](https://github.com/airbytehq/airbyte/actions/runs/5663742866)

File: airbyte-ci/connectors/pipelines/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/README.md:38]: I think that after you run pipx install the `airbyte-ci` the CLI is not directly available in the PATH.
I suggest to add something like: `Open a new terminal session or source your bash profile to make `airbyte-ci` available in your path.`

**Comment by alafanechere** [comment]
@bnchrch I pushed a gitignore change to ignore the legacy pipeline reports paths. Otherwise a `git add .` will stage all the old report file a developer might have locally.

**Comment by alafanechere** [comment]
@bnchrch I updated the poetry lock following the test dependency declaration.

**Pull request merged**